### PR TITLE
Data Export - ResultStorageEngine Interface implementation 

### PR DESCRIPTION
**Pull request by chandrakaran**
Resolves #1272 

## Description
i) Add the ResultStorageEngine Interface.
ii) Add the default implementation of ResultStorageEngine interface with the methods in elide-async.
iii) Create the ResultStore Table.
iv) Store the results in the new ResultStore table in Process Query.
v) Unit Test Case update/IT test update.

## Motivation and Context
This change allows the implementation  of DefaultResultStorageEngine and the helps store the results in the asyncResultStorageEngine.

## How Has This Been Tested?
DefaultResultStorageEngineTest.java has been created for testing
There are 4 tests inside DefaultResultStorageEngineTest.java to test the 4 methods which are implemented in DefaultResultStorageEngine.

## License
I confirm that this contribution is made under an Apache 2.0 license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: elide-async/src/main/java/com/yahoo/elide/async/service/AsyncQueryThread.java [hunk 1]
```diff
@@ -5,6 +5,7 @@
  */
 package com.yahoo.elide.async.service;
 
+
```
Comment (moizarafat) [inline elide-async/src/main/java/com/yahoo/elide/async/service/AsyncQueryThread.java:8]: please remove extra line

File: elide-async/src/main/java/com/yahoo/elide/async/models/AsyncQueryResultStorage.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+import com.yahoo.elide.annotation.Exclude;
+import lombok.Data;
+
+import java.sql.Blob;
+import java.util.UUID;
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.Id;
+
+
```
Comment (moizarafat) [inline elide-async/src/main/java/com/yahoo/elide/async/models/AsyncQueryResultStorage.java:18]: plz remove extra line.

File: elide-async/src/main/java/com/yahoo/elide/async/service/AsyncQueryThread.java [hunk 3]
```diff
@@ -59,6 +63,7 @@ public AsyncQueryThread(AsyncQuery queryObj, User user, Elide elide, QueryRunner
         this.runner = runner;
         this.asyncQueryDao = asyncQueryDao;
         this.apiVersion = apiVersion;
+       // this.resultStorageEngine = resultStorageEngine;
```
Comment (moizarafat) [inline elide-async/src/main/java/com/yahoo/elide/async/service/AsyncQueryThread.java:66]: plz remove commented line.

File: elide-async/src/main/java/com/yahoo/elide/async/service/DefaultResultStorageEngine.java [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+    }
+
+    @Override
+    public AsyncQueryResultStorage storeResults(UUID asyncQueryID, String responseBody) {
+        byte[] temp = responseBody.getBytes();
+        Blob response = null;
+        try {
+            response = new SerialBlob(temp);
+        } catch (SQLException e) {
+            log.error("Exception: {}", e);
```
Comment (moizarafat) [inline elide-async/src/main/java/com/yahoo/elide/async/service/DefaultResultStorageEngine.java:67]: throw the exception back as an illegal state exception.

File: elide-async/src/main/java/com/yahoo/elide/async/service/DefaultResultStorageEngine.java [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+        asyncQueryResultStorage.setId(asyncQueryID);
+        asyncQueryResultStorage.setResult(response);
+
+        URL url = null;
+        try {
+            url = new URL("https://elide.io/asyncQueryID");
+        } catch (MalformedURLException e) {
+            log.error("Exception: {}", e);
+        }
+        return asyncQueryResultStorage;
```
Comment (moizarafat) [inline elide-async/src/main/java/com/yahoo/elide/async/service/DefaultResultStorageEngine.java:79]: return url instead.

File: elide-async/src/main/java/com/yahoo/elide/async/service/DefaultResultStorageEngine.java [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+        } catch (SQLException e) {
+            log.error("Exception: {}", e);
+        }
+        AsyncQueryResultStorage asyncQueryResultStorage = new AsyncQueryResultStorage();
+        asyncQueryResultStorage.setId(asyncQueryID);
+        asyncQueryResultStorage.setResult(response);
+
+        URL url = null;
+        try {
+            url = new URL("https://elide.io/asyncQueryID");
```
Comment (moizarafat) [inline elide-async/src/main/java/com/yahoo/elide/async/service/DefaultResultStorageEngine.java:75]: url should be constructed using three parts base_url, "/asyncQueryResultStore/" , asyncQueryID. base_url can be initialized by setting in the constructor. For elide standalone, baseUrl can be another Settings and in spring we can pass it in application.yaml

File: elide-async/src/main/java/com/yahoo/elide/async/models/AsyncQueryResultStorage.java (hunk 2 above)
Comment (chandrakaran) [reply elide-async/src/main/java/com/yahoo/elide/async/models/AsyncQueryResultStorage.java:18]: done

File: elide-async/src/main/java/com/yahoo/elide/async/service/AsyncQueryThread.java (hunk 3 above)
Comment (chandrakaran) [reply elide-async/src/main/java/com/yahoo/elide/async/service/AsyncQueryThread.java:68]: done

File: elide-async/src/main/java/com/yahoo/elide/async/models/AsyncQueryResultStorage.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+@Exclude //We don't want this to be exposed through the standard Elide APIs.
+@Entity
+@Data
+public class AsyncQueryResultStorage {
+    @Id
+    @Column(columnDefinition = "varchar(36)")
+    private UUID id; //Matches UUID in query.
```
[the remark this answers is not in the document]
Comment (chandrakaran) [reply elide-async/src/main/java/com/yahoo/elide/async/models/AsyncQueryResultStorage.java:25]: changed to String

File: elide-async/src/main/java/com/yahoo/elide/async/service/ResultStorageEngine.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+public interface ResultStorageEngine {
+
+    /**
+     * Stores the result of the query in the AsyncQueryResultStorage table and returns the URL to result.
+     * @param asyncQueryID is the query ID of the AsyncQuery
+     * @param responseBody is the result obtained by running the query
+     * @return it returns the URL from where we can access the result of the query
+     */
+    public AsyncQueryResultStorage storeResults(UUID asyncQueryID, String responseBody);
```
[the remark this answers is not in the document]
Comment (chandrakaran) [reply elide-async/src/main/java/com/yahoo/elide/async/service/ResultStorageEngine.java:22]: changed to String

File: elide-async/src/main/java/com/yahoo/elide/async/service/DefaultResultStorageEngine.java [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.UUID;
+
+import javax.inject.Singleton;
+import javax.sql.rowset.serial.SerialBlob;
+import javax.ws.rs.core.MultivaluedHashMap;
+import javax.ws.rs.core.MultivaluedMap;
+
+
```
[the remark this answers is not in the document]
Comment (chandrakaran) [reply elide-async/src/main/java/com/yahoo/elide/async/service/DefaultResultStorageEngine.java:39]: done

File: elide-async/src/main/java/com/yahoo/elide/async/service/ResultStorageEngine.java [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+ */
+
+package com.yahoo.elide.async.service;
+
+import com.yahoo.elide.async.models.AsyncQueryResultStorage;
+
+import java.util.Collection;
+import java.util.UUID;
+
+public interface ResultStorageEngine {
```
[the remark this answers is not in the document]
Comment (chandrakaran) [reply elide-async/src/main/java/com/yahoo/elide/async/service/ResultStorageEngine.java:14]: Done

File: elide-async/src/main/java/com/yahoo/elide/async/service/DefaultResultStorageEngine.java [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+@Singleton
+@Slf4j
+@Getter
+public class DefaultResultStorageEngine implements ResultStorageEngine {
+
+    @Setter private Elide elide;
+    @Setter private DataStore dataStore;
+    private EntityDictionary dictionary;
+    private RSQLFilterDialect filterParser;
```
[the remark this answers is not in the document]
Comment (chandrakaran) [reply elide-async/src/main/java/com/yahoo/elide/async/service/DefaultResultStorageEngine.java:48]: removed entity dictionary  and filter

File: elide-async/src/main/java/com/yahoo/elide/async/service/DefaultResultStorageEngine.java [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+
+
+@Singleton
+@Slf4j
+@Getter
+public class DefaultResultStorageEngine implements ResultStorageEngine {
+
+    @Setter private Elide elide;
+    @Setter private DataStore dataStore;
+    private EntityDictionary dictionary;
```
[the remark this answers is not in the document]
Comment (chandrakaran) [reply elide-async/src/main/java/com/yahoo/elide/async/service/DefaultResultStorageEngine.java:47]: removed entity dictionary  and filter

File: elide-async/src/main/java/com/yahoo/elide/async/service/AsyncQueryThread.java [hunk 13]
```diff
@@ -95,12 +110,83 @@ else if (queryObj.getQueryType().equals(QueryType.GRAPHQL_V1_0)) {
         queryResultObj.setHttpStatus(response.getResponseCode());
         queryResultObj.setResponseBody(response.getBody());
         queryResultObj.setContentLength(response.getBody().length());
+        queryResultObj.setRecordCount(recCount);
         queryResultObj.setResultType(ResultType.EMBEDDED);
         queryResultObj.setCompletedOn(new Date());
 
+        resultStorageEngine = new DefaultResultStorageEngine();
```
[the remark this answers is not in the document]
Comment (chandrakaran) [reply elide-async/src/main/java/com/yahoo/elide/async/service/AsyncQueryThread.java:121]: Done

File: elide-integration-tests/src/test/java/com/yahoo/elide/async/integration/tests/AsyncIT.java [hunk 14]
```diff
@@ -324,6 +324,7 @@ public void jsonApiRequestTests() throws InterruptedException {
                         .body("data.id", equalTo("ba31ca4e-ed8f-4be0-a0f3-12088fa9263d"))
                         .body("data.type", equalTo("asyncQuery"))
                         .body("data.attributes.result.contentLength", notNullValue())
+                        .body("data.attributes.result.recordCount", equalTo(3))
```
Comment (moizarafat) [inline elide-integration-tests/src/test/java/com/yahoo/elide/async/integration/tests/AsyncIT.java:327]: We should test the ResultStorageEngine table too in AsyncIT.

**Pull request closed without merging**